app/modals/revoke-delegation/revoke-delegation-modal.tsx

```diff
@@ -0,0 +1,288 @@
+
+  const { encryptedMnemonic, salt, publicKey, poxInfo, coreNodeInfo, balance } = useSelector(
+    (state: RootState) => ({
+      salt: selectSalt(state),
+      encryptedMnemonic: selectEncryptedMnemonic(state),
+      publicKey: selectPublicKey(state),
+      poxInfo: selectPoxInfo(state),
+      coreNodeInfo: selectCoreNodeInfo(state),
+      balance: selectAddressBalance(state),
+    })
+  );
+
+  const initialStep = whenWallet({
+    software: RevokeDelegationModalStep.DecryptWalletAndSend,
+    ledger: RevokeDelegationModalStep.SignWithLedgerAndSend,
+  });
+
+  const [step, setStep] = useState(initialStep);
+
+  const createSoftwareWalletTx = useCallback(async (): Promise<StacksTransaction> => {
+    if (!password || !encryptedMnemonic || !salt || !poxInfo || !balance) {
+      throw new Error('One of `password`, `encryptedMnemonic` or `salt` is missing');
+    }
+    if (coreNodeInfo === null) throw new Error('Stacking requires coreNodeInfo');
+
+    const { privateKey } = await decryptSoftwareWallet({
+      ciphertextMnemonic: encryptedMnemonic,
+      salt,
+      password,
+    });
+
+    const txOptions = stackingClient.getRevokeDelegateStxOptions(poxInfo.contract_id);
+    const tx = await makeContractCall({ ...txOptions, senderKey: privateKey });
+    const signer = new TransactionSigner(tx);
+    signer.signOrigin(createStacksPrivateKey(privateKey));
+    return tx;
+  }, [password, encryptedMnemonic, salt, poxInfo, balance, coreNodeInfo, stackingClient]);
+
+  const createLedgerWalletTx = useCallback(
+    async (options: { publicKey: Buffer }): Promise<StacksTransaction> => {
+      if (coreNodeInfo === null) throw new Error('Stacking requires coreNodeInfo');
+      if (!blockstackApp || !poxInfo || !balance)
+        throw new Error('`poxInfo` or `blockstackApp` is not defined');
+
+      const txOptions = stackingClient.getRevokeDelegateStxOptions(poxInfo.contract_id);
+
+      const unsignedTx = await makeUnsignedContractCall({
+        ...txOptions,
+        publicKey: options.publicKey.toString('hex'),
+      });
+
+      const resp: ResponseSign = await blockstackApp.sign(
+        STX_DERIVATION_PATH,
+        unsignedTx.serialize()
+      );
+      if (resp.returnCode !== LedgerError.NoErrors) {
+        throw new Error('Ledger responded with errors');
+      }
+      return unsignedTx.createTxWithSignature(resp.signatureVRS);
+    },
+    [coreNodeInfo, blockstackApp, poxInfo, balance, stackingClient]
+  );
+
+  const broadcastTx = async () => {
+    if (balance === null) return;
+
+    const broadcastActions: Omit<BroadcastTransactionArgs, 'transaction'> = {
+      onBroadcastSuccess: () => {
+        dispatch(homeActions.closeRevokeDelegationModal());
+        setTimeout(() => void mutate('delegation-status'), 2000);
+      },
+      onBroadcastFail: () => setStep(RevokeDelegationModalStep.FailedContractCall),
+    };
+
+    setHasSubmitted(true);
+    await delay(100);
```

Comment:
Can you make a helper, or comment when these delays happen to give some context for their purpose?